### PR TITLE
[CLI] Shorthand notation to instantiate optimizers and lr schedulers [2/3]

### LINKED ISSUE #7576: Easier way to configure optimizers and schedulers in the CLI
## 🚀 Feature
Right now if you want to work with multiple optimizers and/or learning rate schedulers you need to write a whole bunch of ugly boilerplate.  I suggest that the LightningCLI in one way or another to enable that sort of optimizer configuration code in a clean way, which would take care of most use cases.

### Motivation

Right now I have the following code in `configure_optimizers` callback:
```python
    # Configuration.  Add more for learning schedulers, etc.?
    def configure_optimizers(self):
        if self.hparams.optimizer == "Adam":
            optimizer = torch.optim.Adam(
                self.parameters(),
                lr=self.hparams.learning_rate,
                weight_decay=self.hparams.weight_decay,
            )
        elif self.hparams.optimizer == "SGD":
            # Left out the momentum options for now
            optimizer = torch.optim.SGD(
                self.parameters(),
                lr=self.hparams.learning_rate,
                weight_decay=self.hparams.weight_decay,
            )
        elif self.hparams.optimizer == "LBFGS":
            optimizer = torch.optim.LBFGS(
                self.parameters(),
                # or can have self.hparams.learning_rate with warning if too low.
                lr=1,
                tolerance_grad=1e-5,  # can add to parameters if useful.
                tolerance_change=1e-9,  # can add to parameters if useful.
            )
        else:
            print("Invalid optimizer.  See --help")
            sys.exit()

        if self.hparams.lr_scheduler == None:
            return optimizer

        # Setup the scheduler
        scheduler = None
        try:
            if self.hparams.lr_scheduler == "StepLR":
                step_size = self.hparams.StepLR_step_size
                gamma = self.hparams.StepLR_gamma
                scheduler = torch.optim.lr_scheduler.StepLR(
                    optimizer, step_size, gamma)
            elif self.hparams.lr_scheduler == "ReduceLROnPlateau":
                factor = self.hparams.lr_factor
                patience = self.hparams.lr_patience
                scheduler = torch.optim.lr_scheduler.ReduceLROnPlateau(
                    optimizer, mode="min", factor=factor, patience=patience
                )
                return {
                    "optimizer": optimizer,
                    "lr_scheduler": scheduler,
                    "monitor": self.hparams.LRScheduler_metric,
                }
            elif self.hparams.lr_scheduler == "OneCycleLR":
                max_lr = self.hparams.OneCycleLR_max_lr
                epochs = self.hparams.OneCycleLR_epochs
                steps_per_epoch = self.hparams.train_trajectories * (
                    self.hparams.T + 1
                )  # I think this is the same as len(train_data_loader)? #TODO
                scheduler = torch.optim.lr_scheduler.OneCycleLR(
                    optimizer,
                    max_lr=max_lr,
                    epochs=epochs,
                    steps_per_epoch=steps_per_epoch,
                )
        except:
            print("Invalid scheduler configuration.  See --help")
            raise
        return [optimizer], [scheduler]
```
And the commandline configuration looks like
```python
        parser.add_argument("--learning_rate", type=float, default=5e-3)
        parser.add_argument("--LRScheduler_metric",
                            type=str, default="val_loss")
        parser.add_argument(
            "--lr_scheduler",
            default="ReduceLROnPlateau",
            help="Learning rate scheduler name, it can be StepLR, ReduceLROnPlateau, or, OneCycleLR. Defaults to None.",
        )
        parser.add_argument(
            "--optimizer",
            type=str,
            default="Adam",
            help="Choice optimizer: Adam, SGD, or LBFGS.",
        )
        parser.add_argument(
            "--StepLR_step_size",
            type=int,
            default=30,  # TODO there was no actual default in the source code
            help="Step size for StepLR scheduler. Defaults to 30.",
        )
        parser.add_argument(
            "--StepLR_gamma",
            type=float,
            default=0.1,
            help="Gamma for StepLR scheduler. Defaults to 0.1 ",
        )
        # plateau based learning rate
        parser.add_argument(
            "--lr_factor",
            type=float,
            default=0.1,
            help="Factor by which the learning rate will be reduced. new_lr = lr * factor.",
        )
        parser.add_argument(
            "--lr_patience",
            type=int,
            default=5,
            help="Number of epochs with no improvement after which learning rate will be reduced. For example, if patience = 2, then we will ignore the first 2 epochs with no improvement, and will only decrease the LR after the 3rd epoch if the loss still hasn’t improved then",
        )
        parser.add_argument(
            "--OneCycleLR_max_lr",
            type=float,  # TODO no default on the source code
            default=0.1,
            help="Upper learning rate boundary in the cycle. Note that it cannot be a list",
        )
        parser.add_argument(
            "--OneCycleLR_epochs",
            type=int,
            default=10,
            help="The number of epochs to train for. This is used along with steps_per_epoch in order to infer the total number of steps in the cycle if a value for total_steps is not provided. Default: None",
        )
```
or maybe even more.... 

### Pitch

Not exactly sure how you are planning to do recursive configuration stuff, but I can imagine the CLI having a baseline `configure_optimizers` on its own so that the user doesn't need to write it for simple patterns.  Then maybe have defaults in the CLI arguments to have things like
```python
python train.py --model.learning_rate 1e-4 --model.lr_scheduler.type ReduceLROnPlateau --model.lr_scheduler.factor 0.1
```
and
```python
python train.py --model.learning_rate 1e-4 --model.lr_scheduler.type ReduceLROnPlateau --model.lr_scheduler.factor 0.1 --model.optimizer.type Adam --model.optimizer.weight_decay 1e-5
```
or whatever.   And the user just wouldn't implement a `configure_optimizers` or at least would have something simple to call.

Writing this out as a configuration, I have in mind that the user could have something like
```yaml
model:
  decoder_layers:
  - 2
  - 4
  encoder_layers: 12
optimizer:
  Adam:
    weight_decay: 1e-5
lr_scheduler:
   ReduceLROnPlateau:
     factor: 0.1     
trainer:
  accelerator: null
  accumulate_grad_batches: 1
  amp_backend: native
  amp_level: O2
```
or maybe instead:
```yaml
model:
  decoder_layers:
  - 2
  - 4
  encoder_layers: 12
  optimizer:
    Adam:
      weight_decay: 1e-5
  lr_scheduler:
     ReduceLROnPlateau:
       factor: 0.1     
trainer:
  accelerator: null
  accumulate_grad_batches: 1
  amp_backend: native
  amp_level: O2
```

etc.

### Alternatives

See the above monstrosity of what is done now.

### PR DESCRIPTION
**Pull request by carmocca**
## What does this PR do?

Fixes #7576
Implements the feature request in jsonargparse: https://github.com/omni-us/jsonargparse/issues/84
Docs: https://129938-178626720-gh.circle-artifacts.com/0/html/common/lightning_cli.html#optimizers-and-learning-rate-scheduler

See the docs changes for more context.

### Does your PR introduce any breaking changes? If yes, please list them.

None

## Before submitting

- [x] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [x] Did you make sure to **update the documentation** with your changes? (if necessary)
- [x] Did you write any **new necessary tests**? (not for typos and docs)
- [x] Did you verify new and **existing tests pass** locally with your changes?
- [x] Did you list all the **breaking changes** introduced by this pull request?
- [x] Did you **update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)**? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review

- [x] Is this pull request ready for review? (if not, please submit in draft mode)
- [x] Check that all items from **Before submitting** are resolved
- [x] Make sure the title is self-explanatory and the description concisely explains the PR
- [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

### CODE REVIEW

**Review by tchaton — APPROVED** [review]
Love it !

File: docs/source/common/lightning_cli.rst [hunk 1]
```diff
@@ -665,69 +665,131 @@ Optimizers and learning rate schedulers
 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
 
 Optimizers and learning rate schedulers can also be made configurable. The most common case is when a model only has a
-single optimizer and optionally a single learning rate scheduler. In this case the model's
-:class:`~pytorch_lightning.core.lightning.LightningModule` could be left without implementing the
-:code:`configure_optimizers` method since it is normally always the same and just adds boilerplate. The following code
-snippet shows how to implement it:
+single optimizer and optionally a single learning rate scheduler. In this case, the model's
+:meth:`~pytorch_lightning.core.lightning.LightningModule.configure_optimizers` could be left unimplemented since it is
+normally always the same and just adds boilerplate.
```
Comment (awaelchli) [inline docs/source/common/lightning_cli.rst:670]: to me, calling configure_optimizers boilerplate is just outrageous.

File: docs/source/common/lightning_cli.rst [hunk 2]
```diff
@@ -665,69 +665,131 @@ Optimizers and learning rate schedulers
...
+    from pytorch_lightning.utilities.cli import OPTIMIZER_REGISTRY, LR_SCHEDULER_REGISTRY
+
+
+    @OPTIMIZER_REGISTRY
+    class CustomAdam(torch.optim.Adam):
+        ...
+
+
+    @LR_SCHEDULER_REGISTRY
+    class CustomCosineAnnealingLR(torch.optim.lr_scheduler.CosineAnnealingLR):
```
Comment (awaelchli) [inline docs/source/common/lightning_cli.rst:715]: can you give a recommendation what to do if I want to register a optimizer or scheduler from another library?

Comment (awaelchli) [reply docs/source/common/lightning_cli.rst:715]: `OPTIMIZER_REGISTRY.register_package(torch.optim, Optimizer)` ?

**Review by awaelchli — APPROVED** [review]

File: docs/source/common/lightning_cli.rst (hunk 1 above)
Comment (carmocca) [reply docs/source/common/lightning_cli.rst:670]: not everybody thinks that: #7576

File: docs/source/common/lightning_cli.rst (hunk 2 above)
Comment (carmocca) [reply docs/source/common/lightning_cli.rst:715]: That's right, sir.

Added to the docs

**Review by SeanNaren — APPROVED** [review]
Not super familiar with the LightningCLI codebase but looks good!

File: requirements/extra.txt [hunk 3]
```diff
@@ -7,6 +7,6 @@ torchtext>=0.7
 onnx>=1.7.0
 onnxruntime>=1.3.0
 hydra-core>=1.0
-jsonargparse[signatures]>=3.19.0
+jsonargparse[signatures]>=3.19.3
```
Comment (daniellepintz) [inline requirements/extra.txt:10]: why was this change made?

Comment (carmocca) [reply requirements/extra.txt:10]: This PR requires a bugfix upstream to work:

https://github.com/omni-us/jsonargparse/pull/90

The patch was released with `3.19.3`
